File: frontend/src/api/ledgers.ts
// The dynamic-ledger API: the company's own record — goals, decisions, and
// whatever axis this workspace declared — read and written through the host's
// `…/ledgers` routes.
//
// The console does not know what a ledger holds. It is told: every read carries
// the ledger's own `fields`, `statuses` and `sections`, and the UI renders a
// form and a table from that rather than from anything hard-coded here. That is
// the whole point — a ledger an agent declared this morning renders correctly
// this afternoon with no console release.

import type { OpenCompanyClient } from "./client";

/** Where a ledger's rows come from. */
export type LedgerSource =
  /** An append-only log this surface writes. */
  | "events"
  /**
   * A ledger the runtime already owns and renders — the task board. Readable
   * here and written where it lives, which `writtenBy` names.
   */
  | "native";

/** What a field is for, which is how the console renders and sorts it. */
export type FieldRole =
  | "id"
  | "title"
  | "status"
  | "prose"
  | "refs"
  | "owner"
  | "date"
  | "number";

export interface LedgerField {
  name: string;
  role: FieldRole;
  /** A line saying what belongs here, shown under the input. */
  description?: string;
  required?: boolean;
}

export interface LedgerStatus {
  name: string;
  /**
   * What a person reads, when it differs from the wire word.
   *
   * Absent for a company-declared ledger, whose statuses are already written to
   * be read; present on the board, whose columns are `in_progress` and
   * `in_review`. Carrying it on the wire is what let the console delete its own
   * copy of the board's label table.
   */
  label?: string;
  /** Whether this status ends a row's life. A closed row is kept, not deleted. */
  closed?: boolean;
  /** Whether closing into it demands a reason. The host refuses without one. */
  needsReason?: boolean;
}

export interface LedgerSection {
  heading: string;
  blurb?: string;
  statuses?: string[];
  cap: number;
  order: "recorded" | "recent";
}

/** One ledger, as the console lists and renders it. */
export interface LedgerSummary {
  slug: string;
  title: string;
  purpose: string;
  source: LedgerSource;
  /** Where its rendered file lives in the workspace, e.g. `derived/GOALS.md`. */
  derived: string;
  /**
   * How it is actually written, in a sentence. Rendered in place of a compose
   * box on a native ledger — the board is written by the board, and offering a
   * form here would produce a save the host refuses.
   */
  writtenBy: string;
  /** Whether the runtime ships it. A built-in cannot be retired. */
  builtin: boolean;
  fields: LedgerField[];
  statuses: LedgerStatus[];
  sections: LedgerSection[];
  writers?: string[];
  open: number;
  closed: number;
}

/** One row. `fields` is whatever the ledger declares — never a fixed shape. */
export interface LedgerEntry {
  id: string;
  fields: Record<string, string>;
  status: string;
  title: string;
  /** Whether the row's status is one the ledger calls closed. */
  closed: boolean;
  openedBy: LedgerAuthor;
  updatedBy: LedgerAuthor;
  openedAt: number;
  updatedAt: number;
  /** How many writes fold into this row. */
  events: number;
}

export interface LedgerAuthor {
  kind: "human" | "agent" | "system";
  id?: string;
  label?: string;
}

export interface LedgerList {
  ledgers: LedgerSummary[];
  /**
   * Declarations that could not be loaded, and why. Rendered rather than
   * dropped: a company whose ledger stopped appearing has no other way to find
   * out what happened to it.
   */
  faults?: string[];
  /** How many more this company may declare. */
  remaining: number;
}

export interface LedgerRead {
  ledger: LedgerSummary;
  entries: LedgerEntry[];
  /**
   * How many matched before the bound was applied. Shown whenever it exceeds
   * what came back, so a short list is never mistaken for all of them.
   */
  matched: number;
  faults?: string[];
}

export interface LedgerQuery {
  entry?: string;
  status?: string;
  q?: string;
  sort?: "recorded" | "recent";
  limit?: number;
}

/**
 * The body every write sends. There is one write: opening, amending and closing
 * a row all merge fields into it, so reusing an `id` updates that row.
 *
 * A field set to `null` clears it — the one thing a merge cannot otherwise say.
 */
export interface RecordEntryBody {
  id: string;
  fields?: Record<string, string | null>;
  status?: string;
  reason?: string;
}

export async function listLedgers(
  client: OpenCompanyClient,
  company: string,
): Promise<LedgerList> {
  return client.get<LedgerList>(`${client.scopeFor(company)}/ledgers`);
}

export async function readLedger(
  client: OpenCompanyClient,
  company: string,
  slug: string,
  query: LedgerQuery = {},
): Promise<LedgerRead> {
  const params = new URLSearchParams();
  if (query.entry) params.set("entry", query.entry);
  if (query.status) params.set("status", query.status);
  if (query.q) params.set("q", query.q);
  if (query.sort) params.set("sort", query.sort);
  if (query.limit != null) params.set("limit", String(query.limit));
  const suffix = params.toString() ? `?${params.toString()}` : "";
  return client.get<LedgerRead>(
    `${client.scopeFor(company)}/ledgers/${encodeURIComponent(slug)}${suffix}`,
  );
}

/**
 * The exact Markdown `derived/<NAME>.md` holds.
 *
 * Served from the derivation rather than by reading the file back, so a
 * workspace write that failed can never show a stale page.
 */
export async function renderedLedger(
  client: OpenCompanyClient,
  company: string,
  slug: string,
): Promise<string> {
  const { text } = await client.getDocument(
    `${client.scopeFor(company)}/ledgers/${encodeURIComponent(slug)}/rendered`,
  );
  return text;
}

export async function recordEntry(
  client: OpenCompanyClient,
  company: string,
  slug: string,
  body: RecordEntryBody,
): Promise<LedgerEntry> {
  return client.post<LedgerEntry>(
    `${client.scopeFor(company)}/ledgers/${encodeURIComponent(slug)}/entries`,
    body,
  );
}

/**
 * Deletes a row and everything recorded against it.
 *
 * The one irreversible operation on a ledger, and the one an agent cannot
 * reach: teammates close rows, which keeps the reason. So this is only ever
 * called from a person's click.
 */
export async function deleteEntry(
  client: OpenCompanyClient,
  company: string,
  slug: string,
  id: string,
): Promise<void> {
  await client.del<void>(
    `${client.scopeFor(company)}/ledgers/${encodeURIComponent(slug)}/entries/${encodeURIComponent(id)}`,
  );
}

export async function defineLedger(
  client: OpenCompanyClient,
  company: string,
  declaration: unknown,
): Promise<LedgerSummary> {
  return client.post<LedgerSummary>(
    `${client.scopeFor(company)}/ledgers`,
    declaration,
  );
}

/**
 * Retires a declared ledger.
 *
 * `purge` deletes the rows too, and defaults off: retiring a ledger nobody
 * reads is worth doing, and deleting what it recorded is a separate decision
 * that should be made on purpose.
 */
export async function retireLedger(
  client: OpenCompanyClient,
  company: string,
  slug: string,
  purge = false,
): Promise<void> {
  const suffix = purge ? "?purge=true" : "";
  await client.del<void>(
    `${client.scopeFor(company)}/ledgers/${encodeURIComponent(slug)}${suffix}`,
  );
}

/** The field carrying a row's identity, if the ledger declares one. */
export function idField(ledger: LedgerSummary): LedgerField | undefined {
  return ledger.fields.find((field) => field.role === "id");
}

/** The field carrying a row's one-line summary. */
export function titleField(ledger: LedgerSummary): LedgerField | undefined {
  return ledger.fields.find((field) => field.role === "title");
}

/** The field carrying a row's status. */
export function statusField(ledger: LedgerSummary): LedgerField | undefined {
  return ledger.fields.find((field) => field.role === "status");
}

/**
 * The fields a compose form should offer: everything but the id and the status,
 * which the form asks for separately.
 */
export function composableFields(ledger: LedgerSummary): LedgerField[] {
  return ledger.fields.filter(
    (field) => field.role !== "id" && field.role !== "status",
  );
}

/**
 * The sentinel the status filter stores for "do not filter".
 *
 * A `Select` item cannot carry an empty string, so the no-filter choice needs a
 * value of its own — the same constraint the workflow destination picker met
 * with `__none__` in issue #813.
 */
export const EVERY_STATUS = "all";

/**
 * What the status filter should *show* for a stored value (issue #1217).
 *
 * base-ui renders the raw stored value in a collapsed `Select` unless it is
 * given explicit text, so without this the trigger read `all` while its own open
 * list read "Every status", and a chosen board column read `in_progress` while
 * the columns beside it read "In progress". Exactly the leak `destinationLabel`
 * exists to stop on the workflows side.
 *
 * Prefers the declared `label` over the wire `name`, which is what that field is
 * for: the board's statuses carry both, and the console deleted its own copy of
 * that table when the host started sending it. An unrecognized value falls back
 * to itself rather than to the sentinel — a filter we cannot name is still a
 * filter, and rendering it as "Every status" would claim the opposite.
 */
export function statusFilterLabel(
  ledger: LedgerSummary | null | undefined,
  value: string,
): string {
  if (value === EVERY_STATUS || value === "") return "Every status";
  const declared = ledger?.statuses.find((status) => status.name === value);
  if (!declared) return value;
  return `${declared.label ?? declared.name}${declared.closed ? " (closed)" : ""}`;
}

/**
 * Why this ledger is showing nothing — the filtered reading, or `null` when no
 * filter is on and "nothing recorded yet" is the honest answer (issue #1217).
 *
 * A ledger read is filtered **server-side**, so zero rows back means "no
 * matches" and "no rows" indistinguishably. The view knows which it is, because
 * the query and the status are its own state — this is that knowledge written
 * down once, so the empty state stops telling an operator with 14 rows that
 * their ledger is empty while the nav beside it counts them.
 *
 * Pure, and separate from the component, for the reason `workflowSavedToast` is:
 * the branch is the whole bug and it should be assertable without a render.
 */
export function filteredEmptyNotice(
  ledger: LedgerSummary | null | undefined,
  query: string,
  statusFilter: string,
): string | null {
  const searching = query.trim() !== "";
  const narrowed = statusFilter !== EVERY_STATUS && statusFilter !== "";
  if (!searching && !narrowed) return null;
  if (searching && narrowed) {
    return `No rows match “${query.trim()}” with status ${statusFilterLabel(ledger, statusFilter)}.`;
  }
  if (searching) return `No rows match “${query.trim()}”.`;
  return `No rows with status ${statusFilterLabel(ledger, statusFilter)}.`;
}

/**
 * The compose dialog's title (issue #1264).
 *
 * The id field is editable in every non-closing dialog — "Record" opens it
 * empty, "Amend" opens it prefilled but still editable — so whether this says
 * "Amend" cannot be "is `composing.id` non-empty" (that flips true on the
 * first keystroke of a row that has never been saved, before it exists). It
 * has to ask whether the *current* typed id names a row this ledger already
 * has, checked against whatever is currently loaded (`existingIds`) the same
 * way `save()` treats the id — trimmed.
 */
export function composeDialogTitle(
  ledger: LedgerSummary,
  composing: { id: string; closing: boolean },
  existingIds: ReadonlySet<string>,
): string {
  if (composing.closing) return `Close ${composing.id}`;
  return existingIds.has(composing.id.trim())
    ? `Amend ${composing.id}`
    : `New row on ${ledger.title}`;
}

/** The compose dialog's description — the same "does this id exist" read as the title. */
export function composeDialogDescription(
  composing: { id: string },
  existingIds: ReadonlySet<string>,
): string {
  return existingIds.has(composing.id.trim())
    ? "Only what you change is written; everything else on the row is left alone."
    : "Give it a short, readable id — it is how anybody names this row later.";
}

/** Whether this status ends a row's life on this ledger. */
export function isClosingStatus(ledger: LedgerSummary, status: string): boolean {
  return ledger.statuses.some(
    (declared) =>
      declared.name.toLowerCase() === status.trim().toLowerCase() &&
      declared.closed === true,
  );
}

/**
 * Whether closing into `status` demands a reason.
 *
 * Read here so the form can *ask* for one, rather than letting the save fail
 * and reporting the host's refusal — the same rule, met before it bites.
 */
export function statusNeedsReason(
  ledger: LedgerSummary,
  status: string,
): boolean {
  return ledger.statuses.some(
    (declared) =>
      declared.name.toLowerCase() === status.trim().toLowerCase() &&
      declared.needsReason === true,
  );
}

/** Whether this ledger takes writes from this surface at all. */
export function isWritable(ledger: LedgerSummary): boolean {
  return ledger.source === "events";
}

/** A row's byline, for the "recorded by" column. */
export function byline(author: LedgerAuthor): string {
  const name = author.label?.trim() || author.id?.trim();
  return name ? `${name} (${author.kind})` : author.kind;
}

File: src/server/ops/ledgers.rs
//! The ledger surface: `GET/POST /ledgers`, `GET/DELETE /ledgers/{slug}`,
//! `GET/POST /ledgers/{slug}/entries`, `DELETE /ledgers/{slug}/entries/{id}`,
//! and `GET /ledgers/{slug}/rendered`, under both scope forms.
//!
//! # Where the deletion rule is actually enforced
//!
//! [`ScopedCompany::actor`] is `Some` for a signed-in person and `None` for the
//! machine principal — the same distinction the operator-message journal draws.
//! That is what [`author`] turns into a [`LedgerAuthor`], and
//! [`crate::company::ledgers`] refuses a delete from anything but a person.
//!
//! The check therefore lives **once**, in the service, and this module's job is
//! only to name the caller honestly. A route that decided for itself would be a
//! second answer to the question, and the agent tools would need a third.

use axum::extract::{Path, Query};
use axum::http::StatusCode;
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use serde::{Deserialize, Serialize};

use crate::AppState;
use crate::company::ledgers;
use crate::ledger::{Entry, LedgerAuthor, LedgerSource, LedgerSpec};
use crate::ports::types::ActorKind;
use crate::server::error::ApiError;
use crate::server::ops::{ScopedCompany, scoped};

/// Builds the ledger route fragment.
pub fn router() -> Router<AppState> {
    scoped("/ledgers", get(list_ledgers).post(define_ledger))
        .merge(scoped(
            "/ledgers/{slug}",
            get(read_ledger).delete(retire_ledger),
        ))
        // The rendered Markdown the `derived/` folder holds, served from the
        // same derivation rather than by reading the file back — so the console
        // cannot be shown a stale copy if a workspace write ever failed.
        .merge(scoped("/ledgers/{slug}/rendered", get(rendered)))
        .merge(scoped(
            "/ledgers/{slug}/entries",
            post(record_entry).get(read_ledger),
        ))
        .merge(scoped(
            "/ledgers/{slug}/entries/{entry_id}",
            delete(delete_entry),
        ))
}

/// One status, wire-shaped for the console (issue #1266).
///
/// `crate::ledger::StatusSpec` stays plain snake_case in both directions,
/// because it is also what a declared ledger round-trips through in every
/// store backend — a rename there would make a save write one key and a
/// reload expect another, silently dropping `needs_reason` back to `false`.
/// This DTO exists so the API-output shape can be camelCase (matching every
/// other field `LedgerSummary` sends) without that type doing double duty.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct LedgerStatusDto {
    name: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    label: String,
    closed: bool,
    needs_reason: bool,
}

impl From<&crate::ledger::StatusSpec> for LedgerStatusDto {
    fn from(status: &crate::ledger::StatusSpec) -> Self {
        Self {
            name: status.name.clone(),
            label: status.label.clone(),
            closed: status.closed,
            needs_reason: status.needs_reason,
        }
    }
}

/// One ledger as the console lists it.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct LedgerSummary {
    slug: String,
    title: String,
    purpose: String,
    /// `events` or `native`. A native ledger's rows are written by the surface
    /// that owns them, which is why the console renders it read-only here.
    source: LedgerSource,
    /// Where its rendered file lives in the workspace.
    derived: String,
    /// How it is actually written, in a sentence — what the console shows in
    /// place of a compose box on a native ledger.
    written_by: String,
    /// Whether the runtime ships it. A built-in cannot be retired.
    builtin: bool,
    fields: Vec<crate::ledger::Field>,
    statuses: Vec<LedgerStatusDto>,
    sections: Vec<crate::ledger::Section>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    writers: Vec<String>,
    /// Rows not in a closed status.
    open: usize,
    /// Rows in one.
    closed: usize,
}

/// What `GET /ledgers` answers.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct LedgerList {
    ledgers: Vec<LedgerSummary>,
    /// Declarations that could not be loaded, and why.
    ///
    /// Surfaced rather than swallowed: a company whose ledger silently stopped
    /// appearing has no way to find out why, and the answer is always in a
    /// declaration it wrote.
    #[serde(skip_serializing_if = "Vec::is_empty")]
    faults: Vec<String>,
    /// How many more this company may declare.
    remaining: usize,
}

/// One row, as the console renders it.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct EntryRow {
    id: String,
    fields: std::collections::BTreeMap<String, String>,
    status: String,
    title: String,
    /// Whether the row's status is one the ledger calls closed.
    closed: bool,
    opened_by: LedgerAuthor,
    updated_by: LedgerAuthor,
    opened_at: u64,
    updated_at: u64,
    events: u32,
}

impl EntryRow {
    fn of(entry: &Entry, spec: &LedgerSpec) -> Self {
        let status = entry.status(spec);
        Self {
            id: entry.id.clone(),
            fields: entry.fields.clone(),
            closed: spec.is_closed(&status),
            status,
            title: entry.title(spec).to_string(),
            opened_by: entry.opened_by.clone(),
            updated_by: entry.updated_by.clone(),
            opened_at: entry.opened_at_millis,
            updated_at: entry.updated_at_millis,
            events: entry.events,
        }
    }
}

/// What `GET /ledgers/{slug}` answers.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct LedgerRead {
    ledger: LedgerSummary,
    entries: Vec<EntryRow>,
    /// How many matched before the bound was applied — so a short list is
    /// distinguishable from all of them.
    matched: usize,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    faults: Vec<String>,
}

/// The narrowing `GET /ledgers/{slug}` accepts.
#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ReadQuery {
    entry: Option<String>,
    status: Option<String>,
    q: Option<String>,
    sort: Option<String>,
    limit: Option<usize>,
}

/// The body `POST /ledgers/{slug}/entries` accepts.
///
/// One shape for every write, because there is one write: opening, amending and
/// closing a row all merge fields into it. `status` and `reason` are named
/// separately only because a console form has fields for them; both are folded
/// into `fields` before the service sees them, so nothing downstream has two
/// ways of being told the same thing.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RecordBody {
    id: String,
    #[serde(default)]
    fields: std::collections::BTreeMap<String, Option<String>>,
    #[serde(default)]
    status: Option<String>,
    #[serde(default)]
    reason: Option<String>,
}

/// The query `DELETE /ledgers/{slug}` accepts.
#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
struct RetireQuery {
    /// Delete the rows too. Off by default: retiring a ledger nobody reads is
    /// worth doing, and deleting what it recorded is a separate decision.
    #[serde(default)]
    purge: bool,
}

/// The ledger context for this request's company.
fn ctx(scope: &ScopedCompany) -> ledgers::Ledgers {
    scope.runtime.as_ref().into()
}

/// The caller, named honestly.
///
/// A signed-in person becomes a [`LedgerAuthor::human`]; the machine principal
/// becomes a system author, which the service refuses every deletion from. That
/// is deliberate and not an oversight: the platform credential is a tenant, not
/// a person, and *only a person deletes* has to mean a person.
async fn author(scope: &ScopedCompany) -> LedgerAuthor {
    match &scope.actor {
        Some(actor) if matches!(actor.kind, ActorKind::Operator | ActorKind::User) => {
            LedgerAuthor::human(actor.id.clone(), human_label(scope, &actor.id).await)
        }
        Some(actor) => LedgerAuthor::agent(actor.id.clone()),
        None => LedgerAuthor::system("platform"),
    }
}

/// A person's byline, resolved through the same user store the admin-only
/// `GET …/users` route reads (`src/server/users/admin.rs::list_users`) — but
/// called directly here, with no `require_admin` gate, so every signed-in
/// member sees a real name rather than only admins.
///
/// Falls back to the raw id on any lookup failure — a deleted user, a store
/// error — because a display-name lookup must never fail the write it is
/// merely decorating.
async fn human_label(scope: &ScopedCompany, actor_id: &str) -> String {
    match scope
        .runtime
        .users()
        .get_user(scope.runtime.id(), actor_id)
        .await
    {
        Ok(Some(user)) => {
            let display_name = user.display_name.as_deref().unwrap_or("").trim();
            if display_name.is_empty() {
                user.email
            } else {
                display_name.to_string()
            }
        }
        Ok(None) | Err(_) => actor_id.to_string(),
    }
}

async fn summary(ctx: &ledgers::Ledgers, spec: &LedgerSpec) -> Result<LedgerSummary, ApiError> {
    let entries = ledgers::entries(ctx, spec).await?;
    Ok(LedgerSummary {
        slug: spec.slug.clone(),
        title: spec.title.clone(),
        purpose: spec.purpose.clone(),
        source: spec.source,
        derived: spec.derived.clone(),
        written_by: spec.written_by.clone(),
        builtin: spec.builtin,
        fields: spec.fields.clone(),
        statuses: spec.statuses.iter().map(LedgerStatusDto::from).collect(),
        sections: spec.sections.clone(),
        writers: spec.writers.clone(),
        open: entries.open_count(spec),
        closed: entries.closed_count(spec),
    })
}

async fn list_ledgers(scope: ScopedCompany) -> Result<Json<LedgerList>, ApiError> {
    let registry = ledgers::registry(&ctx(&scope)).await?;
    let mut out = Vec::new();
    for spec in registry.specs() {
        out.push(summary(&ctx(&scope), spec).await?);
    }
    let declared = registry.specs().iter().filter(|spec| !spec.builtin).count();
    Ok(Json(LedgerList {
        ledgers: out,
        faults: registry.faults().to_vec(),
        remaining: crate::ledger::MAX_DECLARED.saturating_sub(declared),
    }))
}

async fn define_ledger(
    scope: ScopedCompany,
    Json(document): Json<serde_json::Value>,
) -> Result<(StatusCode, Json<LedgerSummary>), ApiError> {
    let spec = ledgers::define(&ctx(&scope), &document).await?;
    let summary = summary(&ctx(&scope), &spec).await?;
    Ok((StatusCode::CREATED, Json(summary)))
}

async fn read_ledger(
    scope: ScopedCompany,
    Path(path): Path<LedgerPath>,
    Query(query): Query<ReadQuery>,
) -> Result<Json<LedgerRead>, ApiError> {
    let registry = ledgers::registry(&ctx(&scope)).await?;
    let spec = registry.require(&path.slug)?;
    let read = ledgers::read(
        &ctx(&scope),
        spec,
        &ledgers::Query {
            entry: query.entry,
            status: query.status,
            text: query.q,
            sort: query.sort,
            limit: query.limit,
        },
    )
    .await?;
    Ok(Json(LedgerRead {
        ledger: summary(&ctx(&scope), spec).await?,
        entries: read.entries.iter().map(|e| EntryRow::of(e, spec)).collect(),
        matched: read.matched,
        faults: read.faults,
    }))
}

/// The exact Markdown `derived/<NAME>.md` holds.
async fn rendered(scope: ScopedCompany, Path(path): Path<LedgerPath>) -> Result<String, ApiError> {
    let registry = ledgers::registry(&ctx(&scope)).await?;
    let spec = registry.require(&path.slug)?;
    Ok(ledgers::render(&ctx(&scope), spec).await?)
}

async fn record_entry(
    scope: ScopedCompany,
    Path(path): Path<LedgerPath>,
    Json(body): Json<RecordBody>,
) -> Result<Json<EntryRow>, ApiError> {
    let registry = ledgers::registry(&ctx(&scope)).await?;
    let spec = registry.require(&path.slug)?;
    let mut fields = body.fields;
    if let Some(status) = body.status
        && let Some(field) = spec.status_field()
    {
        fields.insert(field.name.clone(), Some(status));
    }
    if let Some(reason) = body.reason {
        fields.insert(crate::ledger::REASON_FIELD.to_string(), Some(reason));
    }
    let entry =
        ledgers::record(&ctx(&scope), spec, &author(&scope).await, &body.id, fields).await?;
    Ok(Json(EntryRow::of(&entry, spec)))
}

async fn delete_entry(
    scope: ScopedCompany,
    Path(path): Path<EntryPath>,
) -> Result<StatusCode, ApiError> {
    let registry = ledgers::registry(&ctx(&scope)).await?;
    let spec = registry.require(&path.slug)?;
    let removed =
        ledgers::delete_entry(&ctx(&scope), spec, &author(&scope).await, &path.entry_id).await?;
    Ok(if removed {
        StatusCode::NO_CONTENT
    } else {
        StatusCode::NOT_FOUND
    })
}

async fn retire_ledger(
    scope: ScopedCompany,
    Path(path): Path<LedgerPath>,
    Query(query): Query<RetireQuery>,
) -> Result<StatusCode, ApiError> {
    ledgers::retire(&ctx(&scope), &author(&scope).await, &path.slug, query.purge).await?;
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Debug, Deserialize)]
struct LedgerPath {
    // The scoped forms differ in whether they carry `{id}`, so the path struct
    // has to tolerate both. `slug` is the only segment either form guarantees.
    #[serde(default)]
    #[allow(dead_code)]
    id: Option<String>,
    slug: String,
}

#[derive(Debug, Deserialize)]
struct EntryPath {
    #[serde(default)]
    #[allow(dead_code)]
    id: Option<String>,
    slug: String,
    entry_id: String,
}

#[cfg(test)]
#[path = "ledgers_test.rs"]
mod test;
